### ISSUE TITLE
LoadWith() in a CTE

### ISSUE THREAD

**Issue by azhdari** [linq2db/linq2db#3407]
Hello,

I try to implement a utility function to return `Paged Data` and `Total Count`.  

```c#
IQueryable<T> mainQuery = ...;
IQueryable<T> query = mainQuery;
string sortBy = ...;
int skip = 0;
int take = 10;

var cte = query.AsCte();
var data = await cte.
    OrderBy(p => Sql.Property<object>(p, sortBy)).
    Skip(skip).
    Take(take).
    ToListAsync(token);
var count = await cte.CountAsync(token);
return (data, count);
```
Generated query (This is the actual generated query, with model details):
```sql
--  PostgreSQL.9.5 PostgreSQL (asynchronously)
DECLARE @take Integer -- Int32
SET     @take = 10
DECLARE @skip Integer -- Int32
SET     @skip = 0

WITH cte
(
        "CommentId",
        "Id",
        "CreatedBy",
        "CreatedAt"
)
AS
(
        SELECT
                t1."CommentId",
                t1."Id",
                t1."CreatedBy",
                t1."CreatedAt"
        FROM
                "CommentLikes" t1
)
SELECT
        t2."Id",
        t2."CommentId",
        t2."CreatedBy",
        t2."CreatedAt"
FROM
        cte t2
ORDER BY
        t2."CommentId"
LIMIT :take OFFSET :skip 
---
WITH cte ("Id")
AS
(
        SELECT
                t1."Id"
        FROM
                "CommentLikes" t1
)
SELECT
        Count(*)
FROM
        cte t2
```
It works most of the time, not when we use `LoadWith()` inside a query though.
```c#
IQueryable<T> query = mainQuery.LoadWith(x => x.User);
```
It will generate this query
```sql
--  PostgreSQL.9.5 PostgreSQL (asynchronously)
DECLARE @take Integer -- Int32
SET     @take = 10
DECLARE @skip Integer -- Int32
SET     @skip = 0

WITH cte ("CommentId")
AS
(
        SELECT
                t1."CommentId",
                t1."Id",
                t1."CreatedBy",
                t1."CreatedAt",
                "a_User"."Id",
                "a_User"."FirstName",
                "a_User"."LastName",
                "a_User"."DisplayName",
                "a_User"."CreatedAt",
                "a_User"."CreatedBy"
        FROM
                "CommentLikes" t1
                        INNER JOIN "Users" "a_User" ON t1."CreatedBy" = "a_User"."Id"
)
SELECT
        1
FROM
        cte t2
ORDER BY
        t2."CommentId"
LIMIT :take OFFSET :skip 

```
It throws an error:
```
Exception message: Column must be between 0 and 0
Stack trace: System.IndexOutOfRangeException: Column must be between 0 and 0
   at Npgsql.NpgsqlDataReader.GetFieldType(Int32 ordinal)
   at LinqToDB.DataProvider.DataProviderBase.GetReaderExpression(IDataReader reader, Int32 idx, Expression readerExpression, Type toType)
   at LinqToDB.Data.DataConnection.LinqToDB.IDataContext.GetReaderExpression(IDataReader reader, Int32 idx, Expression readerExpression, Type toType)
   at LinqToDB.Expressions.ConvertFromDataReaderExpression.GetColumnReader(IDataContext dataContext, MappingSchema mappingSchema, IDataReader dataReader, Type type, IValueConverter converter, Int32 idx, Expression dataReaderExpr, Boolean forceNullCheck)
   at LinqToDB.Expressions.ConvertFromDataReaderExpression.Reduce(IDataContext dataContext, IDataReader dataReader, Expression dataReaderParam)
   at LinqToDB.Linq.QueryRunner.Mapper`1.<>c.<TransformMapperExpression>b__5_1(TransformMapperExpressionContext context, Expression e)
   at LinqToDB.Expressions.TransformVisitor`1.Transform(Expression expr)
...
 at LinqToDB.Linq.ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken)
   at LinqToDB.AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token)
   at LanguageExt.TaskExtensions.<>c__DisplayClass31_0`1.<<ToAff>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at LanguageExt.Thunks.ThunkAsync`1.<>c__DisplayClass5_0.<<Lazy>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at LanguageExt.Thunks.ThunkAsync`1.Eval()
[11:22:14 INF] Executing JsonResult, writing value of type '<>f__AnonymousType0`2[[System.Int32, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]'.
...
```

Models:
```c#
[Table("CommentLikes")]
public record LikeEntity(
    [property: Column(SkipOnInsert = true, IsPrimaryKey = true)]
    Guid Id,
    [property: Column] Guid CommentId,
    [property: Column] Guid CreatedBy,
    [property: Column] DateTimeOffset CreatedAt,
    [property: Association(ThisKey = nameof(LikeEntity.CreatedBy), OtherKey = nameof(UserEntity.Id), CanBeNull = false)]
    UserEntity? User
);
[Table("Users")]
public record UserEntity(
    [property: Column(SkipOnInsert = true, IsPrimaryKey = true)]
    Guid Id,
    [property: Column] string? FirstName,
    [property: Column] string? LastName,
    [property: Column] string DisplayName,
    [property: Column] DateTimeOffset CreatedAt,
    [property: Column] Guid CreatedBy
);
```

### Environment details
linq2db version: *3.6.0*
Database: *PostgreSQL*
Operating system: *DB: Ubuntu (Docker), Dev: MacOS*
.NET: *6*

**Comment by sdanyliv** [comment]
Will check, anyway there is better solution: https://github.com/linq2db/linq2db/issues/1402#issuecomment-438961009

**Comment by azhdari** [comment]
Thank you for the solution.

> not sure how it will work with LoadWith

I'll check it.

**Comment by azhdari** [comment]
> not sure how it will work with LoadWith

Thank you, it works perfectly, the issue of LoadWith in CTE still exists though.

**Comment by MaceWindu** [comment]
Implemented in `6.0.0-preview.1`
